src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -337,15 +337,22 @@ def setup_impl(self, exclude_pool):
         mntinfo = self.middleware.call_sync('filesystem.mount_info')
         if config['pool'] != boot_pool:
             if not any(filter_list(mntinfo, [['mount_source', '=', config['pool']]])):
-                ds = self.middleware.call_sync('zfs.dataset.query', [['id', '=', config['basename']]])
+                ds = self.middleware.call_sync(
+                    'zfs.resource.query_impl',
+                    {'paths': [config['basename']], 'properties': ['encryption']}
+                )
                 if not ds:
                     # Pool is not mounted (e.g. HA node B), temporary set up system dataset on the boot pool
                     msg = 'Root dataset for pool %r is not available, and dataset %r does not exist, '
                     msg += 'temporarily setting up system dataset on boot pool'
                     self.logger.debug(msg, config['pool'], config['basename'])
                     self.force_pool = boot_pool
                     config = self.middleware.call_sync('systemdataset.config')
-                elif ds[0]['encrypted'] and ds[0]['locked'] and ds[0]['key_format']['value'] != 'PASSPHRASE':
+                elif (
+                    ds[0]['properties']['encryption']['raw'] != 'off'
+                    and ds[0]['properties']['keystatus']['raw'] != 'available'
+                    and ds[0]['properties']['keyformat']['raw'] != 'passphrase'
```

Comment:
Nothing wrong, but a couple of minor nits:

- Prior code was a _little_ clearer.  Maybe add a comment here.
- Having strings like `'available'` and `'passphrase'` scattered through the code invites typo bugs.  :-/